File: go/policy.go
package main

import "math"

type Tri int

const (
	TriFalse Tri = iota
	TriTrue
	TriUncertain
)

func (t Tri) MarshalJSON() ([]byte, error) {
	switch t {
	case TriTrue:
		return []byte("true"), nil
	case TriFalse:
		return []byte("false"), nil
	default:
		return []byte(`"uncertain"`), nil
	}
}

type Policy struct {
	SwitchConfidence float64
	InterruptAt      float64
	UncertainMargin  float64
}

var defaultPolicy = Policy{SwitchConfidence: 0.72, InterruptAt: 0.75, UncertainMargin: 0.12}

type Decision struct {
	Disposition string
	Executing   string
	Because     string
}

type TargetCandidate struct {
	ID       string
	Relation string
	Distance *float64
	Health   *float64
	Kind     string
}

type Bound struct {
	TargetID     *string
	TargetSource string
}

func resolvePolicy(in map[string]any) (Policy, error) {
	policy := defaultPolicy
	if in == nil {
		return policy, nil
	}
	for _, key := range []string{"switchConfidence", "interruptAt", "uncertainMargin"} {
		raw, ok := in[key]
		if !ok || raw == nil {
			continue
		}
		value, ok := asFloat(raw)
		if !ok || math.IsNaN(value) || value < 0 || value > 1 {
			return Policy{}, errBad(400, "policy."+key+" 必须是 0 到 1 之间的数字")
		}
		switch key {
		case "switchConfidence":
			policy.SwitchConfidence = value
		case "interruptAt":
			policy.InterruptAt = value
		case "uncertainMargin":
			policy.UncertainMargin = value
		}
	}
	return policy, nil
}

func classifyBoolean(probability *float64, margin float64) Tri {
	if probability == nil || math.IsNaN(*probability) {
		return TriUncertain
	}
	if *probability >= 0.5+margin {
		return TriTrue
	}
	if *probability <= 0.5-margin {
		return TriFalse
	}
	return TriUncertain
}

func classifyInterrupt(probability *float64, interruptAt float64) Tri {
	if probability == nil || math.IsNaN(*probability) {
		return TriUncertain
	}
	if *probability >= interruptAt {
		return TriTrue
	}
	if *probability <= 1-interruptAt {
		return TriFalse
	}
	return TriUncertain
}

func decideDisposition(current, suggested string, confidence *float64, interrupt Tri, holdKey *string, switchConfidence float64) Decision {
	current = trim(current)
	expanded := suggested
	if holdKey != nil && suggested == *holdKey && current != "" {
		expanded = current
	}
	if current == "" {
		return Decision{Disposition: "switch", Executing: expanded, Because: "first-decision"}
	}
	if expanded == current {
		return Decision{Disposition: "continue", Executing: current, Because: "still-fitting"}
	}
	if interrupt == TriTrue {
		return Decision{Disposition: "switch", Executing: expanded, Because: "interrupt"}
	}
	if confidence == nil || *confidence >= switchConfidence {
		return Decision{Disposition: "switch", Executing: expanded, Because: "confident"}
	}
	return Decision{Disposition: "hold", Executing: current, Because: "hysteresis"}
}

func tacticNeedsTarget(tactic string) bool {
	switch tactic {
	case "hold", "patrol", "hide", "flee":
		return false
	default:
		return true
	}
}

func bindTarget(tactic, modelTargetID string, roster []TargetCandidate) Bound {
	if !tacticNeedsTarget(tactic) {
		return Bound{TargetSource: "none"}
	}
	if modelTargetID != "none" && modelTargetID != "" {
		for _, entry := range roster {
			if entry.ID == modelTargetID {
				id := modelTargetID
				return Bound{TargetID: &id, TargetSource: "model"}
			}
		}
	}
	if fallback := fallbackTarget(tactic, roster); fallback != "" {
		id := fallback
		return Bound{TargetID: &id, TargetSource: "geometric-fallback"}
	}
	return Bound{TargetSource: "none"}
}

func scoreBand(score float64, labels []string) ScoreRead {
	max := 0
	if n := len(labels) - 1; n > 0 {
		max = n
	}
	clamped := math.Min(float64(max), math.Max(0, score))
	level := int(math.Round(clamped))
	if level < 0 {
		level = 0
	}
	if level > max {
		level = max
	}
	label := "unknown"
	if level < len(labels) {
		label = labels[level]
	}
	return ScoreRead{Score: math.Round(clamped*100) / 100, Level: level, Label: label}
}

func fallbackTarget(tactic string, roster []TargetCandidate) string {
	ranked := append([]TargetCandidate(nil), roster...)
	sortByDistance(ranked)
	if tactic == "assist" {
		var allies []TargetCandidate
		for _, entry := range ranked {
			if entry.Relation == "ally" {
				allies = append(allies, entry)
			}
		}
		sortAllies(allies)
		if len(allies) > 0 {
			return allies[0].ID
		}
		return ""
	}
	if tactic == "investigate" || tactic == "interact" {
		for _, entry := range ranked {
			if entry.Kind == "interest" || entry.Kind == "hazard" || entry.Kind == "prop" {
				return entry.ID
			}
		}
		if len(ranked) > 0 {
			return ranked[0].ID
		}
		return ""
	}
	for _, entry := range ranked {
		if entry.Relation == "enemy" {
			return entry.ID
		}
	}
	if len(ranked) > 0 {
		return ranked[0].ID
	}
	return ""
}

func sortByDistance(items []TargetCandidate) {
	for i := 1; i < len(items); i++ {
		j := i
		for j > 0 && distOrFar(items[j]) < distOrFar(items[j-1]) {
			items[j], items[j-1] = items[j-1], items[j]
			j--
		}
	}
}

func sortAllies(items []TargetCandidate) {
	for i := 1; i < len(items); i++ {
		j := i
		for j > 0 && allyLess(items[j], items[j-1]) {
			items[j], items[j-1] = items[j-1], items[j]
			j--
		}
	}
}

func allyLess(a, b TargetCandidate) bool {
	ha, hb := 1.0, 1.0
	if a.Health != nil {
		ha = *a.Health
	}
	if b.Health != nil {
		hb = *b.Health
	}
	if ha != hb {
		return ha < hb
	}
	return distOrFar(a) < distOrFar(b)
}

func distOrFar(item TargetCandidate) float64 {
	if item.Distance == nil {
		return 1e9
	}
	return *item.Distance
}

File: rust/src/policy.rs
use serde::Serialize;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Tri {
    False,
    True,
    Uncertain,
}

impl Serialize for Tri {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        match self {
            Tri::True => serializer.serialize_bool(true),
            Tri::False => serializer.serialize_bool(false),
            Tri::Uncertain => serializer.serialize_str("uncertain"),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct ScoreRead {
    pub score: f64,
    pub level: i32,
    pub label: String,
}

#[derive(Debug, Clone)]
pub struct Policy {
    pub switch_confidence: f64,
    pub interrupt_at: f64,
    pub uncertain_margin: f64,
}

impl Default for Policy {
    fn default() -> Self {
        Self {
            switch_confidence: 0.72,
            interrupt_at: 0.75,
            uncertain_margin: 0.12,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Decision {
    pub disposition: String,
    pub executing: String,
    pub because: String,
}

#[derive(Debug, Clone)]
pub struct TargetCandidate {
    pub id: String,
    pub relation: String,
    pub distance: Option<f64>,
    pub health: Option<f64>,
    pub kind: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Bound {
    pub target_id: Option<String>,
    pub target_source: String,
}

pub fn resolve_policy(
    input: Option<&serde_json::Map<String, serde_json::Value>>,
) -> Result<Policy, crate::error::IntuitionError> {
    let mut policy = Policy::default();
    let Some(input) = input else {
        return Ok(policy);
    };
    for key in ["switchConfidence", "interruptAt", "uncertainMargin"] {
        let Some(raw) = input.get(key) else { continue };
        if raw.is_null() {
            continue;
        }
        let Some(value) = as_f64(raw) else {
            return Err(crate::error::err_bad(
                400,
                format!("policy.{key} 必须是 0 到 1 之间的数字"),
            ));
        };
        if !value.is_finite() || !(0.0..=1.0).contains(&value) {
            return Err(crate::error::err_bad(
                400,
                format!("policy.{key} 必须是 0 到 1 之间的数字"),
            ));
        }
        match key {
            "switchConfidence" => policy.switch_confidence = value,
            "interruptAt" => policy.interrupt_at = value,
            _ => policy.uncertain_margin = value,
        }
    }
    Ok(policy)
}

pub fn classify_boolean(probability: Option<f64>, margin: f64) -> Tri {
    let Some(p) = probability.filter(|n| n.is_finite()) else {
        return Tri::Uncertain;
    };
    if p >= 0.5 + margin {
        Tri::True
    } else if p <= 0.5 - margin {
        Tri::False
    } else {
        Tri::Uncertain
    }
}

pub fn classify_interrupt(probability: Option<f64>, interrupt_at: f64) -> Tri {
    let Some(p) = probability.filter(|n| n.is_finite()) else {
        return Tri::Uncertain;
    };
    if p >= interrupt_at {
        Tri::True
    } else if p <= 1.0 - interrupt_at {
        Tri::False
    } else {
        Tri::Uncertain
    }
}

pub fn decide_disposition(
    current: &str,
    suggested: &str,
    confidence: Option<f64>,
    interrupt: Tri,
    hold_key: Option<&str>,
    switch_confidence: f64,
) -> Decision {
    let current = current.trim();
    let expanded = if hold_key == Some(suggested) && !current.is_empty() {
        current
    } else {
        suggested
    };
    if current.is_empty() {
        return Decision {
            disposition: "switch".into(),
            executing: expanded.into(),
            because: "first-decision".into(),
        };
    }
    if expanded == current {
        return Decision {
            disposition: "continue".into(),
            executing: current.into(),
            because: "still-fitting".into(),
        };
    }
    if interrupt == Tri::True {
        return Decision {
            disposition: "switch".into(),
            executing: expanded.into(),
            because: "interrupt".into(),
        };
    }
    if confidence.is_none_or(|c| c >= switch_confidence) {
        return Decision {
            disposition: "switch".into(),
            executing: expanded.into(),
            because: "confident".into(),
        };
    }
    Decision {
        disposition: "hold".into(),
        executing: current.into(),
        because: "hysteresis".into(),
    }
}

pub fn tactic_needs_target(tactic: &str) -> bool {
    !matches!(tactic, "hold" | "patrol" | "hide" | "flee")
}

pub fn bind_target(tactic: &str, model_target_id: &str, roster: &[TargetCandidate]) -> Bound {
    if !tactic_needs_target(tactic) {
        return Bound {
            target_id: None,
            target_source: "none".into(),
        };
    }
    if model_target_id != "none"
        && !model_target_id.is_empty()
        && roster.iter().any(|e| e.id == model_target_id)
    {
        return Bound {
            target_id: Some(model_target_id.into()),
            target_source: "model".into(),
        };
    }
    if let Some(id) = fallback_target(tactic, roster) {
        return Bound {
            target_id: Some(id),
            target_source: "geometric-fallback".into(),
        };
    }
    Bound {
        target_id: None,
        target_source: "none".into(),
    }
}

pub fn score_band(score: f64, labels: &[String]) -> ScoreRead {
    let max = labels.len().saturating_sub(1) as f64;
    let clamped = score.clamp(0.0, max);
    let mut level = clamped.round() as i32;
    level = level.clamp(0, max as i32);
    let label = labels
        .get(level as usize)
        .cloned()
        .unwrap_or_else(|| "unknown".into());
    ScoreRead {
        score: (clamped * 100.0).round() / 100.0,
        level,
        label,
    }
}

fn fallback_target(tactic: &str, roster: &[TargetCandidate]) -> Option<String> {
    let mut ranked = roster.to_vec();
    ranked.sort_by(|a, b| dist_or_far(a).total_cmp(&dist_or_far(b)));
    if tactic == "assist" {
        let mut allies: Vec<_> = ranked
            .into_iter()
            .filter(|e| e.relation == "ally")
            .collect();
        allies.sort_by(|a, b| {
            let ha = a.health.unwrap_or(1.0);
            let hb = b.health.unwrap_or(1.0);
            ha.total_cmp(&hb)
                .then(dist_or_far(a).total_cmp(&dist_or_far(b)))
        });
        return allies.first().map(|e| e.id.clone());
    }
    if tactic == "investigate" || tactic == "interact" {
        if let Some(found) = ranked
            .iter()
            .find(|e| matches!(e.kind.as_str(), "interest" | "hazard" | "prop"))
        {
            return Some(found.id.clone());
        }
        return ranked.first().map(|e| e.id.clone());
    }
    ranked
        .iter()
        .find(|e| e.relation == "enemy")
        .or_else(|| ranked.first())
        .map(|e| e.id.clone())
}

fn dist_or_far(item: &TargetCandidate) -> f64 {
    item.distance.unwrap_or(1e9)
}

pub fn as_f64(v: &serde_json::Value) -> Option<f64> {
    v.as_f64()
        .or_else(|| v.as_i64().map(|n| n as f64))
        .or_else(|| v.as_u64().map(|n| n as f64))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn low_confidence_holds() {
        let got = decide_disposition(
            "patrol",
            "engage",
            Some(0.4),
            Tri::False,
            Some("hold"),
            0.72,
        );
        assert_eq!(got.disposition, "hold");
        assert_eq!(got.executing, "patrol");
        assert_eq!(got.because, "hysteresis");
    }

    #[test]
    fn interrupt_switches() {
        let got = decide_disposition("patrol", "flee", Some(0.4), Tri::True, Some("hold"), 0.72);
        assert_eq!(got.executing, "flee");
        assert_eq!(got.because, "interrupt");
    }

    #[test]
    fn hold_expands() {
        let got = decide_disposition("patrol", "hold", Some(0.95), Tri::True, Some("hold"), 0.72);
        assert_eq!(got.executing, "patrol");
        assert_eq!(got.because, "still-fitting");
    }

    #[test]
    fn bind_skips_patrol() {
        assert_eq!(bind_target("patrol", "player", &[]).target_source, "none");
    }

    #[test]
    fn assist_weakest_ally() {
        let got = bind_target(
            "assist",
            "none",
            &[
                TargetCandidate {
                    id: "wolf".into(),
                    relation: "enemy".into(),
                    distance: Some(2.0),
                    health: Some(1.0),
                    kind: "creature".into(),
                },
                TargetCandidate {
                    id: "mira".into(),
                    relation: "ally".into(),
                    distance: Some(6.0),
                    health: Some(0.2),
                    kind: "npc".into(),
                },
                TargetCandidate {
                    id: "squire".into(),
                    relation: "ally".into(),
                    distance: Some(3.0),
                    health: Some(0.9),
                    kind: "npc".into(),
                },
            ],
        );
        assert_eq!(got.target_id.as_deref(), Some("mira"));
        assert_eq!(got.target_source, "geometric-fallback");
    }
}
